src/lib/add-on/clients/add-on_auto.rb

```diff
@@ -1,328 +1,319 @@
-              if Builtins.haskey(prod, "name")
-                name = Ops.get_string(prod, "name", "")
-                Builtins.y2milestone("Preferred name: %1", name)
-                # Or use the one returned by Pkg::RepositoryScan
-              else
-                repos_at_url = Pkg.RepositoryScan(Pkg.ExpandedUrl(media))
-                # [ ["Product Name", "Path" ] ]
-                Builtins.foreach(repos_at_url) do |one_repo|
-                  if Ops.get(one_repo, 1, "") == pth
-                    name = Ops.get(one_repo, 0, "")
-                    raise Break
-                  end
-                end
-                Builtins.y2milestone("Preferred name: %1", name)
-              end
-
-              Ops.set(repos, [found_at, "name"], name)
-              Ops.set(repos, [found_at, "priority"], prod["priority"]) if prod.key?("priority")
-              Pkg.SourceEditSet(repos)
-            end
-          end
-          if Ops.get_string(prod, "product", "") != ""
-            Builtins.y2milestone(
-              "Installing product: %1",
-              Ops.get_string(prod, "product", "")
-            )
-            Pkg.ResolvableInstall(Ops.get_string(prod, "product", ""), :product)
-          else
-            Builtins.y2warning("No product to install")
-          end
-        end
 
-        # reread agents, redraw wizard steps, etc.
-        AddOnProduct.ReIntegrateFromScratch
-
-        @ret = true
-      # Reads configuration of add-ons from the current system
-      # to memory. To get that configuration, use Export() functionality.
-      #
-      # @return [Boolean]
-      elsif @func == "Read"
-        if !PackageLock.Check
-          Builtins.y2error("Cannot get package lock")
-          return false
+      AddOnProduct.Import("add_on_products" => valid_add_on_products)
+    end
+
+    def summary
+      formatted_add_on = AddOnProduct.add_on_products.map do |add_on|
+        "<li>" \
+          "Media: #{add_on["media_url"]}, " \
+          "Path: #{add_on["product_dir"]}, " \
+          "Product: #{add_on["product"]}" \
+        "</li>"
+      end
+
+      ["<ul>", formatted_add_on, "</ul>"].flatten.join("\n")
+    end
+
+    def modified?
+      AddOnProduct.modified
+    end
+
+    def modified
+      AddOnProduct.modified = true
+    end
+
+    def reset
+      AddOnProduct.add_on_products = []
+
+      {}
+    end
+
+    def change
+      Wizard.CreateDialog
+      AutoinstSoftware.pmInit
+      PackageCallbacks.InitPackageCallbacks
+
+      ret = RunAddOnMainDialog(
+        false,
+        true,
+        true,
+        Label.BackButton,
+        Label.OKButton,
+        Label.CancelButton,
+        false
+      )
+
+      UI.CloseDialog
+
+      ret
+    end
+
+    def export
+      AddOnProduct.Export
+    end
+
+    def write
+      AddOnProduct.add_on_products.each { |add_on| create_source(add_on) }
+
+      # reread agents, redraw wizard steps, etc.
+      AddOnProduct.ReIntegrateFromScratch
+
+      true
```

Comment:
I'm not sure if always returning `true` is a good idea. What about adding some error handling? I guess that `create_source(add_on)` above might fail...

Reply:
Oh I see, `create_source` reports error internally using the `report_error_for` method... But I'm still not sure if always returning `true` is OK. If it is then it should be at least documented here that it's OK and the error is reported elsewhere.

Reply:
Ok, I'll think a little about it and make some changes to improve this.

Thanks

Reply:
@lslezak what about the new approach? (I just updated the PR)